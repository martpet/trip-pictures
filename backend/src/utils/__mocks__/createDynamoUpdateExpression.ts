export const createDynamoUpdateExpression = vi
  .fn()
  .mockName('createDynamoUpdateExpression')
  .mockReturnValue({
    ExpressionAttributeNames: 'dummyExpressionAttributeNames',
    ExpressionAttributeValues: 'dummyExpressionAttributeValues',
    UpdateExpression: 'dummyUpdateExpression',
  });
