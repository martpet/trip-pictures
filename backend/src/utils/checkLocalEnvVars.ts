import { EnvName } from '~/types';

export const checkLocalEnvVars = (envName: EnvName) => {
  const names = [
    'VITE_PERSONAL_SUBDOMAIN',
    'PERSONAL_DEV_ACCOUNT_ID',
    'PERSONAL_DEV_ACCOUNT_REGION',
  ];
  if (envName === 'personal') {
    names.forEach((varName) => {
      if (!process.env[varName]) {
        throw new Error(`environment variable "${varName}" is missing from .env.local`);
      }
    });
  }
};
