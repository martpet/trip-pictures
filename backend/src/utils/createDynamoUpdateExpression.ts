export const createDynamoUpdateExpression = (
  data: Record<string, unknown>,
  parentKey?: string
) => {
  const ExpressionAttributeNames: Record<string, string> = {};
  const ExpressionAttributeValues: Record<string, unknown> = {};
  const actions: string[] = [];

  Object.entries(data).forEach(([attr, val]) => {
    actions.push(`${parentKey ? `#${parentKey}.` : ''}#${attr} = :${attr}`);
    ExpressionAttributeNames[`#${attr}`] = attr;
    ExpressionAttributeValues[`:${attr}`] = val;
    if (parentKey) {
      ExpressionAttributeNames[`#${parentKey}`] = parentKey;
    }
  });

  return {
    UpdateExpression: `set ${actions.join(',')}`,
    ExpressionAttributeNames,
    ExpressionAttributeValues,
  };
};
