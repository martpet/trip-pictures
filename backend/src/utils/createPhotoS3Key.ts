type Props = {
  userId: string;
  fingerprint: string;
};

export function createPhotoS3Key({ userId, fingerprint }: Props) {
  return `${userId}/${fingerprint}`;
}
