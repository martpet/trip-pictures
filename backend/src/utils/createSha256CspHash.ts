import crypto from 'crypto';

export const createSha256CspHash = (content: string) =>
  `sha256-${crypto.createHash('sha256').update(content).digest('base64')}`;
