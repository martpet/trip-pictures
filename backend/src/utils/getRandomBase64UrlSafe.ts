import crypto from 'crypto';
import { promisify } from 'util';

export const getRandomBase64UrlSafe = async (size: number) => {
  const cryptoRandomBytes = promisify(crypto.randomBytes);
  const bytes = await cryptoRandomBytes(size);
  return bytes.toString('base64url');
};
