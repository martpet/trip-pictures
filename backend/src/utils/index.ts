export * from '~/../../shared/utils';
//
export * from './checkLocalEnvVars';
export * from './createDynamoUpdateExpression';
export * from './createPhotoS3Key';
export * from './createSha256CspHash';
export * from './getAllowedOrigins';
export * from './getRandomBase64UrlSafe';
export * from './getRoleCredentials';
export * from './objectValuesToJson';
