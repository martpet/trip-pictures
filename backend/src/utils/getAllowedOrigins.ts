import { appEnvs, localhostUrl } from '~/consts';
import { getCdkEnv } from '~/utils';

export const getAllowedOrigins = () => {
  const envName = getCdkEnv();
  const { appDomain } = appEnvs[envName];

  const origins = [`https://${appDomain}`];

  if (envName === 'personal') {
    origins.push(localhostUrl);
  }

  return origins;
};
