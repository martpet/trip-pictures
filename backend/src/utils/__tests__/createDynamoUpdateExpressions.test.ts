import { itReturns } from '~/constructs/Api/utils';
import { createDynamoUpdateExpression } from '~/utils/createDynamoUpdateExpression';

const args: Parameters<typeof createDynamoUpdateExpression> = [
  {
    param1: 'param1Value',
    param2: 'param2Value',
  },
];

const argsWithParentKey = [...args, 'dummyParentKey'] as unknown as Parameters<
  typeof createDynamoUpdateExpression
>;

describe('createDynamoUpdateExpression', () => {
  itReturns(createDynamoUpdateExpression, args);

  describe('when parent key is provided', () => {
    itReturns(createDynamoUpdateExpression, argsWithParentKey);
  });
});
