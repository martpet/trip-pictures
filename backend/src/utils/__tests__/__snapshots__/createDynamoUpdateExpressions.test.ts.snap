// Vitest Snapshot v1

exports[`createDynamoUpdateExpression > returns a correct value 1`] = `
{
  "ExpressionAttributeNames": {
    "#param1": "param1",
    "#param2": "param2",
  },
  "ExpressionAttributeValues": {
    ":param1": "param1Value",
    ":param2": "param2Value",
  },
  "UpdateExpression": "set #param1 = :param1,#param2 = :param2",
}
`;

exports[`createDynamoUpdateExpression > when parent key is provided > returns a correct value 1`] = `
{
  "ExpressionAttributeNames": {
    "#dummyParentKey": "dummyParentKey",
    "#param1": "param1",
    "#param2": "param2",
  },
  "ExpressionAttributeValues": {
    ":param1": "param1Value",
    ":param2": "param2Value",
  },
  "UpdateExpression": "set #dummyParentKey.#param1 = :param1,#dummyParentKey.#param2 = :param2",
}
`;
